### ISSUE TITLE
[next] build very noisy with gcc 8.1.1

### ISSUE THREAD

**Issue by dschwoerer** [boutproject/BOUT-dev#1252]
Most are from `-Wnull-dereference`

See https://omadesala.alfa1024.info/~bout-test/2018-09-05_17:18/build.log

There are some others as well.

Good that `-Werror` was not included.

**Comment by d7919** [comment]
It looks like most of these are coming from `output.hxx` and are duplicated wherever this is included.

**Comment by ZedThree** [comment]
What's your configure line and/or `BOUT_FLAGS`? I don't see these with gcc 8.2.1

**Comment by bendudson** [comment]
Would changing the `base` member to a reference fix this? 
https://github.com/boutproject/BOUT-dev/blob/next/include/output.hxx#L197
Probably advisable anyway, if we're guaranteeing that `base` is not null. Also, does it need to be public, since there is a `getBase()`?

**Comment by dschwoerer** [comment]
A reference is not working, as the dynamic_cast is the issue.
```
   Output *getBase() {
     if (base_is_cond) {
-      return dynamic_cast<ConditionalOutput *>(base)->getBase();
+      auto basecond = dynamic_cast<ConditionalOutput *>(base);
+      ASSERT0(basecond != nullptr);
+      return basecond->getBase();
     } else {
       return base;
     }
```
The ASSERT0 is required (rather then ASSERT1) as otherwise the compiler does not see the check.

I have a commit that removes dynamic_cast all together, and rather stores the non-casted pointer of basecond - 2a65198bd5b98705c3e53bf21f4bc1ffc1e11a05

I noticed many warnings about unused variables, when checks are disabled.
The code in ConditionalOutput may had an issue as base was public, but it was not suffering from a lacking check after dynamic_cast, as it get's checked before that the cast will be successful.

**Comment by bendudson** [comment]
Fixed in PR #1256
